### PR TITLE
Transliterator data struct generation

### PR DESCRIPTION
**Pull request by skius**
Depends on #3822
Part of #3736 

This PR adds (a first version of) data struct generation. It leaves some aspects for follow-up PRs (notably cursors and legacy ID => internal ID mapping).

### CODE REVIEW

File: experimental/transliterator_parser/src/compile/pass2.rs [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+        ds::Rule {
+            ante: ante.into(),
+            key: key.into(),
+            post: post.into(),
+            replacer: replacer.into(),
+            cursor_offset,
+        }
+    }
+
+    fn compile_single_id(&mut self, id: &parse::SingleId) -> ds::SimpleId<'static> {
```
Comment (robertbastian) [inline experimental/transliterator_parser/src/compile/pass2.rs:269]: this doesn't use `self`, so it doesn't need to be a method. Maybe put it on `parse::SingleId`?

File: experimental/transliterator_parser/src/compile/pass2.rs [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+                .filter
+                .as_ref()
+                .map(|f| f.code_points().clone())
+                .unwrap_or(CodePointInversionList::all()),
+        }
+    }
+
+    // returns the standin char
+    fn compile_variable(&mut self, var: &str) -> char {
+        if let Some(c) = self.var_to_char.get(var) {
```
Comment (robertbastian) [inline experimental/transliterator_parser/src/compile/pass2.rs:294]: nit: prefer matching on refs over dereffing
```suggestion
        if let Some(&c) = self.var_to_char.get(var) {
```

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: experimental/transliterator_parser/src/compile.rs [hunk 3]
```diff
@@ -71,6 +71,13 @@
 // - replace uses of single-element variables with the element itself ($a = [a-z]; $a > a; => [a-z] > a;)
 // - flatten single-element sets into literals ([a] > b; => a > b;)
 
+// TODO: add this optimization note somewhere:
```
[the remark this answers is not in the document]
Comment (skius) [reply experimental/transliterator_parser/src/compile.rs:74]: Done in #3827

File: experimental/transliterator_parser/Cargo.toml [hunk 4]
```diff
@@ -24,13 +24,18 @@ include = [
 [package.metadata.docs.rs]
 all-features = true
 
+[dev-dependencies]
```
[the remark this answers is not in the document]
Comment (skius) [reply experimental/transliterator_parser/Cargo.toml:27]: Doing in #3827

File: experimental/transliterator_parser/src/lib.rs [hunk 5]
```diff
@@ -25,6 +25,8 @@
 )]
 #![warn(missing_docs)]
 
+extern crate core;
```
[the remark this answers is not in the document]
Comment (skius) [reply experimental/transliterator_parser/src/lib.rs:28]: Done in #3827

File: experimental/transliterator_parser/src/compile/pass2.rs (hunk 1 above)
Comment (skius) [reply experimental/transliterator_parser/src/compile/pass2.rs:269]: I prefer not to spread out the compilation over multiple modules, would an associated function here be fine as well?

File: experimental/transliterator_parser/src/lib.rs [hunk 6]
```diff
@@ -112,3 +128,173 @@ where
...
+        # splits up the forward rules
+        forward rule that > splits up rule groups ;
+        :: InterIndic-Devanagari ;
+        "#;
+
+        let (forward, reverse) = parse(source).expect("parsing failed");
+        let forward = forward.expect("forward transliterator expected");
+        let reverse = reverse.expect("reverse transliterator expected");
+
+        {
```
[the remark this answers is not in the document]
Comment (skius) [reply experimental/transliterator_parser/src/lib.rs:169]: Doing in https://github.com/unicode-org/icu4x/pull/3827

File: experimental/transliterator_parser/src/compile/pass2.rs (hunk 1 above)
Comment (skius) [reply experimental/transliterator_parser/src/compile/pass2.rs:269]: Discussed: associated function fine. Doing in https://github.com/unicode-org/icu4x/pull/3827

**Review by robertbastian — APPROVED** [review]

**Pull request merged**